### PR TITLE
apps sc: Update opendistro helm chart to 1.12

### LINKED ISSUE #164: [2] Update opendistro to 1.12.0
**Why?**
- To get access to multi role nodes.
- No need for standalone client, master, and data nodes with our application of elasticsearch
  - Makes more sense to either have data/master/client nodes or master/client + data nodes.

**High prio?**
- Yes

**How?**
- Read relase notes
  - Take actions accordingly
- Test the new version out
  - Install in new cluster
  - Upgrade running cluster
  - Confirm functionality
- Update config to allow for all-in-one statefulset or let the user fully decide the topology.
  - Test upgrade path

### PR DESCRIPTION
**Pull request by OlleLarsson**
**What this PR does / why we need it**:
Upgrades the helm chart which gives us access to creating multirole nodes.

Updates test according to the new settings.

Updates the dev flavor to deploy a **single** multi-role node(kind of just to see if it works well for us or not).

**Which issue this PR fixes**:
fixes #164 

**Special notes for reviewer**:

I've changed the init flow a bit.
The default value for storageclasses for elasticsearch has been changed to `null`, which means that the cluster default will be used. This makes sense imo as it decreases the init-complexity and is in line with most other helm charts.

With the new version there are some repeated deprecation (every 30s) logs in the client nodes
```
[2021-02-10T18:59:20,396][DEPRECATION][o.e.d.c.m.IndexNameExpressionResolver] [opendistro-es-client-85b7d4f6c5-spkvf] this request accesses system indices: [.kibana_1], but in a future major version, direct access to system indices will be prevented by default
```
https://discuss.elastic.co/t/fresh-install-if-elk-7-10-floods-deprecation-log/259972/3
This is not a big deal imo.

The new version of od introduced changes to ism actions which means that the configurer needs additional permissions. This mean that upgrading will require running `securityadmin.sh`. I will add upgrade docs soon.

I'm running a single node in my cluster to see if it works or not
image

With a single node we won't get a "green" cluster state as indicies by default consist of 1 primary and 1 replica shard, and as we only have one data node, the replica shard cannot be assgined anywhere. To me this does not matter as it's just a dev cluster, just thought that iäd mention it.

**Checklist:**

- [ ] Added relevant notes to [WIP-CHANGELOG.md](https://github.com/elastisys/compliantkubernetes-apps/blob/main/WIP-CHANGELOG.md)
- [x] Proper commit message prefix on all commits
- [ ] Updated the [public facing documentation](https://github.com/elastisys/compliantkubernetes)
- Is this changeset backwards compatible for existing clusters? Applying:
    - [ ] is completely transparent, will not impact the workload in any way.
    - [x] requires running a migration script.
    - [ ] will create noticeable cluster degradation.
          E.g. logs or metrics are not being collected or Kubernetes API server
          will not be responding while upgrading.
    - [ ] requires draining and/or replacing nodes.
    - [ ] will change any APIs.
          E.g. removes or changes any CK8S config options or Kubernetes APIs.
    - [ ] will break the cluster.
          I.e. full cluster migration is required.

### CODE REVIEW

File: WIP-CHANGELOG.md [hunk 1]
```diff
@@ -1,4 +1,10 @@
 ### Release notes
+- With the update of the opendistro helm chart you can now decide whether or not you want dedicated deployments for data and client/ingest nodes.
+  By setting `elasticsearch.dataNode.dedicatedPods: false` and `elasticsearch.clientNode.dedicatedPods: false`,
+  The master node statefulset will assume all roles.
```
Comment (lentzi90) [inline WIP-CHANGELOG.md:4]:
```suggestion
  the master node statefulset will assume all roles.
```

File: WIP-CHANGELOG.md [hunk 2]
```diff
@@ -1,4 +1,10 @@
...
+- To get some of the new default values for resource requests on Harbor pods you will first need to remove the resource requests that you have in your Harbor config and then run `ck8s init` to get the new values.
```
Comment (lentzi90) [inline WIP-CHANGELOG.md:5]: Should this be here? It looks unrelated :thinking:

File: migration/v0.9.x-v0.10.x/upgrade-apps.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
+
+# Upgrade v0.9.x to v0.10.0
+
+You will need to follow these steps in order to upgrade each Compliant Kubernetes environment from v0.8.x to v0.9.0. The instructions assume that you work from the root of the `compliantkubernetes-apps` repository, has the latest changes already pulled and configured `CK8S_CONFIG_PATH` correctly.
```
Comment (lentzi90) [inline migration/v0.9.x-v0.10.x/upgrade-apps.md:4]:
```suggestion
You will need to follow these steps in order to upgrade each Compliant Kubernetes environment from v0.9.x to v0.10.0.
The instructions assume that you work from the root of the `compliantkubernetes-apps` repository, has the latest changes already pulled and configured `CK8S_CONFIG_PATH` correctly.
```

**Review by lentzi90 — COMMENTED** [review]
Awesome!

I like the change to the storage config and I agree that the issues you mentioned seem to be no big deal. To get the cluster to green state, would it be possible to set the number of replicas to 0 or something like that?
One thing I noticed is that we go from 1.10 to 1.12 directly with this change. Is this safe or should we first go to 1.11? I haven't checked how it is for opendistro, but I know some projects don't support skipping over minor versions, while others have "odd versions are dev releases". :smile:

File: migration/v0.9.x-v0.10.x/upgrade-apps.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+   # Wait for master pod 0 to be up and for elasticsearch to have started
+
+   # Reload security config
+   ./bin/ck8s ops kubectl sc -n elastic-system exec opendistro-es-master-0 -- chmod +x ./plugins/opendistro_security/tools/securityadmin.sh
+   ./bin/ck8s ops kubectl sc -n elastic-system exec opendistro-es-master-0 -- ./plugins/opendistro_security/tools/securityadmin.sh \
+       -cd plugins/opendistro_security/securityconfig/ \
+       -icl -nhnv \
+       -cacert config/admin-root-ca.pem \
+       -cert config/admin-crt.pem \
+       -key config/admin-key.pem
```
Comment (viktor-f) [inline migration/v0.9.x-v0.10.x/upgrade-apps.md:30]: This will remove all other (manually added) users like it's done before? Perhaps it could be worth adding a note about that

File: migration/v0.9.x-v0.10.x/upgrade-apps.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+5. Upgrade service cluster applications
+  ```bash
+
+   # Upgrade Elasticsearch
+   ./bin/ck8s ops helmfile sc -l app=opendistro apply
+
+   # Wait for master pod 0 to be up and for elasticsearch to have started
+
+   # Reload security config
```
Comment (viktor-f) [inline migration/v0.9.x-v0.10.x/upgrade-apps.md:23]: You need to reload the config in a new terminal right? If so then write that here, I know that I have missed that in the past and waited until the release timed out.

**Review by viktor-f — COMMENTED** [review]
Looks good :) 

Some minor upgrade comments.

File: WIP-CHANGELOG.md (hunk 2 above)
Comment (viktor-f) [reply WIP-CHANGELOG.md:5]: Yes, the harbor line is old.

File: migration/v0.9.x-v0.10.x/upgrade-apps.md (hunk 4 above)
Comment (OlleLarsson) [reply migration/v0.9.x-v0.10.x/upgrade-apps.md]: I've changed it so that only roles.yml is reloaded instead of everything.
I've also added a note.

File: migration/v0.9.x-v0.10.x/upgrade-apps.md (hunk 5 above)
Comment (OlleLarsson) [reply migration/v0.9.x-v0.10.x/upgrade-apps.md]: I've added a note.

**Review by Xartos — APPROVED** [review]
LGTM! Great upgrade for all involved!

File: WIP-CHANGELOG.md (hunk 2 above)
Comment (OlleLarsson) [reply WIP-CHANGELOG.md]: Yes, a copy/paster error from my side.

**Review by lentzi90 — APPROVED** [review]
LGTM!

**Review by viktor-f — APPROVED** [review]
LGTM!

File: migration/v0.9.x-v0.10.x/upgrade-apps.md (hunk 4 above)
Comment (viktor-f) [reply migration/v0.9.x-v0.10.x/upgrade-apps.md]: Nice, so rolebindings will stay as well as users?

File: migration/v0.9.x-v0.10.x/upgrade-apps.md (hunk 5 above)
Comment (viktor-f) [reply migration/v0.9.x-v0.10.x/upgrade-apps.md]: Thanks :slightly_smiling_face:

File: migration/v0.9.x-v0.10.x/upgrade-apps.md (hunk 4 above)
Comment (OlleLarsson) [reply migration/v0.9.x-v0.10.x/upgrade-apps.md]: Yes

**Comment by OlleLarsson** [comment]
> Awesome!
> 
> I like the change to the storage config and I agree that the issues you mentioned seem to be no big deal. To get the cluster to green state, would it be possible to set the number of replicas to 0 or something like that?

Not impossible but it isn't trivial.
It might be possible to create an index template that's applied to all indices that sets `number_of_replicas: 0`.
However, I suspect that there are some problematic (system)indices that will have been created before we can create the required index templates.
It might be possible to configure the aforementioned indices via configuration(elasticsearch.yml) but I haven't looked further into it.

> One thing I noticed is that we go from 1.10 to 1.12 directly with this change. Is this safe or should we first go to 1.11? I haven't checked how it is for opendistro, but I know some projects don't support skipping over minor versions, while others have "odd versions are dev releases".

It should be fine. I've checked the release notes and the upgrade thing on their website and it didn't say anything specific that'd effect us.

**Comment by OlleLarsson** [comment]
Tested upgrading once more, and a fresh install using the defaults from the dev falvor and it seems to work smoothly.
